File: killrvideo-service-search/src/main/java/com/killrvideo/service/search/grpc/SearchServiceGrpcValidator.java
package com.killrvideo.service.search.grpc;

import static com.killrvideo.utils.ValidationUtils.initErrorString;
import static com.killrvideo.utils.ValidationUtils.validate;
import static org.apache.commons.lang3.StringUtils.isBlank;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.util.Assert;

import io.grpc.stub.StreamObserver;
import killrvideo.search.SearchServiceOuterClass.GetQuerySuggestionsRequest;
import killrvideo.search.SearchServiceOuterClass.SearchVideosRequest;

@Component
public class SearchServiceGrpcValidator  {
    private static final Logger LOGGER = LoggerFactory.getLogger(SearchServiceGrpcValidator.class);

    public void validateGrpcRequest_GetQuerySuggestions(GetQuerySuggestionsRequest request, StreamObserver<?> streamObserver) {
        final StringBuilder errorMessage = initErrorString(request);
        boolean isValid = true;
        if (isBlank(request.getQuery())) {
            errorMessage.append("\t\tquery string should be provided for get video suggestions request\n");
            isValid = false;
        }
        if (request.getPageSize() <= 0) {
            errorMessage.append("\t\tpage size should be strictly positive for get video suggestions request\n");
            isValid = false;
        }
        Assert.isTrue(validate(LOGGER, streamObserver, errorMessage, isValid),
                "Invalid parameter for 'getQuerySuggestions'");
    }
    
    /**
     * Validation for search.
     */
    public void validateGrpcRequest_SearchVideos(SearchVideosRequest request, StreamObserver<?> streamObserver) {
        final StringBuilder errorMessage = initErrorString(request);
        boolean isValid = true;
        if (isBlank(request.getQuery())) {
            errorMessage.append("\t\tquery string should be provided for search videos request\n");
            isValid = false;
        }
        if (request.getPageSize() <= 0) {
            errorMessage.append("\t\tpage size should be strictly positive for search videos request\n");
            isValid = false;
        }
        Assert.isTrue(validate(LOGGER, streamObserver, errorMessage, isValid), "Invalid parameter for 'searchVideos'");
    }
}

File: killrvideo-service-search/src/main/kotlin/com/killrvideo/service/search/grpc/SearchServiceGrpcValidator.kt
package com.killrvideo.service.search.grpc

import com.killrvideo.utils.FluentValidator
import killrvideo.search.SearchServiceOuterClass.GetQuerySuggestionsRequest
import killrvideo.search.SearchServiceOuterClass.SearchVideosRequest
import mu.KotlinLogging
import org.apache.commons.lang3.StringUtils
import org.springframework.stereotype.Component

@Component
class SearchServiceGrpcValidator {
    private val logger = KotlinLogging.logger {  }

    fun validateGrpcRequest_GetQuerySuggestions(request: GetQuerySuggestionsRequest) {
        FluentValidator.of("getQuerySuggestions", request, logger)
            .notEmpty("query string", StringUtils.isBlank(request.query))
            .positive("page size", request.pageSize <= 0)
            .validate()
    }

    /**
     * Validation for search.
     */
    fun validateGrpcRequest_SearchVideos(request: SearchVideosRequest) {
        FluentValidator.of("searchVideos", request, logger)
            .notEmpty("query string", StringUtils.isBlank(request.query))
            .positive("page size", request.pageSize <= 0)
            .validate()
    }
}
